### DISCUSSION TITLE
apply-k3s-agent pending on nodes with taints

### DISCUSSION THREAD

**Discussion by clouedoc** [vitobotta/hetzner-k3s#240] [Q&A]
I get these pods which are pending, on my server groups with taints:

<img width="1182" alt="image" src="https://github.com/vitobotta/hetzner-k3s/assets/13921610/b5ad308a-3e5f-452d-96a2-88f38a663b23">
<img width="699" alt="image" src="https://github.com/vitobotta/hetzner-k3s/assets/13921610/16c8fc68-d4e0-4460-bac3-5f0425571f29">

Am I doing something wrong?

**Comment by vitobotta** [comment]
I converted this to a discussion since it's not an issue with the tool. Please use discussions for generic Kubernetes questions.

You need to configure deployments to _tolerate_ taints otherwise the pods cannot be scheduled on tainted nodes.

**Reply by clouedoc** [reply]
Hi, thanks for your reply.

I believe that this is a problem with the tool given that these pods are from the system-upgrade namespace (e.g. the system-upgrade controller was created, but without the necessary configuration to tolerate the taints).

To fix it, I'm going to convert my taints from NoSchedule and NoExecute to PreferNoSchedule.

I hope this will fix this kind of issue

**Reply by clouedoc** [reply]
PS: I'll make sure to open a discussion next time, if I'm not 100% sure that it's a problem with hetzner-k3s 🙂

**Reply by vitobotta** [reply]
Not sure why you think it's a problem with the tool. If you adapt the configuration to taint some nodes with custom taints, these are taints that e.g. the system upgrade controller is not aware of, and therefore the pods will remain unscheduled if there are no nodes that can satisfy the requirements of those pods. The tool installs the system upgrade controller, the CSI driver and the CCM as they are provided by the respective authors, it doesn't change them. Same thing with the autoscaler, apart from a little configuration to configure the node pools. Therefore if you use taints that are not recognized by e.g. the system upgrade controller, it's expected that its pods will remain unscheduled.

**Reply by clouedoc** [reply]
Here's why: I expected hetzner-k3s to automatically configure the built-in tools to tolerate user taints because I believe that they are critical to the maintenance of the system -- it seemed important to me that these pods get scheduled even on nodes with taints. I was concurrently having random nodes disconnecting from the cluster, which made me suspect these pending pods (turns out that it was just a resource usage issue)

However, I understand your point; I should've expected that using taints might prevent potentially critical workloads to be scheduled on my nodes, _even if I did it through hetzner-k3s itself_.

To sum up: in this specific issue that I encountered; does the responsibility lie in me (_not having read enough about k8s_) or within the tool (_is it reasonable to expect the tolerations to be created by hetzner-k3s when configuring taints through it's configuration file?_)

PS: apologies for the super short original issue message... I now realise that it was not complete enough to convey my perspective

**Reply by vitobotta** [reply]
Like I mentioned the tool installs system components as they are without any modification otherwise it would be more difficult to manage them. If you want to use custom taints just keep a static pool with a single node, even  small one, untainted so those components can be scheduled. Another option is to download the manifests for those components and add tolerations for your tains, and specify the paths of those manifests in the configuration file. v1.1.3 allows you to configure these paths so you can do that if you prefer.

**Reply by clouedoc** [reply]
The issue with these pods is that they cannot schedule onto other nodes; they require to run on node X, Y, Z
So using a single untainted node to run them is not an option
